### ISSUE TITLE
vcd_vapp: Add access control api

### ISSUE THREAD

**Issue by jpbuecken** [vmware/terraform-provider-vcd#509]
Hello,

we need the following enhancement:

Set access control of a vapp via terraform

https://pubs.vmware.com/vca/index.jsp?topic=%2Fcom.vmware.vcloud.api.doc_56%2FGUID-5E62D4D8-5D9A-4251-99E2-029F714B0C3E.html

Use case: We build VMs for our customer or coworker. 
Customer has read access to the org to see his environment, but the environment ist managed by us.
So we need to configure the shared setting for the vapp, such that the customer can see the vapp.

### Affected Resource(s)
Please list the resources as a list, for example:
- vcd_vapp

### Terraform Configuration Files
```hcl
resource "vcd_vapp" "myvapp" {

  name = myvapp
  issharedtoeveryone = true
  everyoneAccessLevel = "ReadOnly"
}

```

### Expected Behavior
Configure Access Control in vapp resource

### Actual Behavior
Not implemented

**Comment by lvirbalas** [comment]
UI analog:

<img width="1167" alt="Screenshot 2020-06-29 at 17 03 38" src="https://user-images.githubusercontent.com/11330689/86015689-8f12fb00-ba2a-11ea-86de-efb3b27bae2b.png">

**Comment by dataclouder** [comment]
Here is how the interface would result:

```hcl
resource "vcd_vapp" "myvapp" {

  name = myvapp
  shared {
    shared_to_everyone = true
    access_level       = "ReadOnly"
  }  
} 
```

```hcl
resource "vcd_vapp" "myvapp" {

  name = myvapp

  shared {
    user_name    = "username1"
    access_level = "ReadWrite"
  }
  shared {
    user_name    = "username2"
    access_level = "FullControl"
  }
}
```

This, however, would be rejected (both all users and specific user)

```hcl
resource "vcd_vapp" "myvapp" {

  name = myvapp

  shared {
    shared_to_everyone = true
    access_level       = "ReadOnly"
  }
  shared {
    user_name    = "username2"
    access_level = "FullControl"
  }
}
```

**Comment by lvirbalas** [comment]
Nice. Few questions:
1. How we define sharing for groups?
2. `shared` or `sharing`? "sharing" seems more clear to me and it's in UI. How it's called in API?

**Comment by dataclouder** [comment]
> Nice. Few questions:
> 
>     1. How we define sharing for groups?

Haven't got to that point yet. Need to investigate. Our support for groups is still Green. (cc @Didainius )

> 
>     2. `shared` or `sharing`? "sharing" seems more clear to me and it's in UI. How it's called in API?

If we want to stay close to the UI, we should use `share`

**Comment by lvirbalas** [comment]
Giuseppe mentioned another wider important aspect:

> One side effect of implementing sharing (called "accessControl" in the API) for vApps is that the same structure and similar calls apply to Catalogs and VDCs.
> https://vdc-repo.vmware.com/vmwb-repository/dcr-public/037ccaee-649a-417e-b365-1331034fb28d/1f0fd9eb-0238-4af6-89b5-7e6636f29c65/doc/doc/types/ControlAccessParamsType.html

We should think whether it's most efficient to implement this for all resources that use this structure in a single batch.

p.s. Here's how Catalog sharing (from right-click menu) looks like:

<img width="872" alt="Screenshot 2020-06-30 at 12 04 26" src="https://user-images.githubusercontent.com/11330689/86107228-3bf08500-baca-11ea-8e88-73cfaedff95e.png">

**Comment by jpbuecken** [comment]
Minutes too late, was writing this:

> 2. `shared` or `sharing`? "sharing" seems more clear to me and it's in UI. How it's called in API?

In API, it is controlAccess  / ControlAccessParams.  
Offtopic but related: Access Control Lists are also applicable for catalogs.

https://pubs.vmware.com/vca/index.jsp?topic=%2Fcom.vmware.vcloud.api.doc_56%2FGUID-A1A86769-7DA6-4577-A940-07ADDE06E250.html

> If we want to stay close to the UI, we should use `share`

Ok for me, maybe API is less likely to change: if you want to stay close to API, maybe `acl` for access control list.

**Comment by dataclouder** [comment]
Regarding names, whenever possible, we try to
* stay close to API in the low level library ([go-vcloud-director](https://github.com/vmware/go-vcloud-director))
* Reuse terminology from the UI in Terraform

**Comment by dataclouder** [comment]
Something that may be necessary to implement this feature: data sources for both org user and org groups.
I will check whether we can define easily without data sources.

**Comment by dataclouder** [comment]
Implemented in PR #543
